### PR TITLE
 [new] Added ipfs as a media source

### LINKED ISSUE #242: [Feature] Add IPFS as a music source
### Describe the problem

Once [feature/settings-ui](https://github.com/Arlodotexe/strix-music/tree/feature/settings-ui) is merged, we'll have IPFS support in the App.

With IPFS comes a reference to `OwlCore.Kubo`, which also contains implementations of `OwlCore.Storage` that work with IPFS, IPNS and MFS. 

### Describe the proposed change

We should use this to add IPFS as a music source in the app.

The user should have these options during setup:
- [Text entry] Add using an IPNS address
- [Text entry] Add using a CID
- [Text entry] Add using an MFS path
- [boolean option] Preload folder contents now
- Import loaded StorageCore into MFS (keep setting control disabled for now, we'll implement later)

It should be tested that the folder is valid and retrievable before allow the user to add the music source.

### Alternatives

_No response_

### Additional info

_No response_

### Help us help you

Yes, I'd like to be assigned to work on this item.

### PR DESCRIPTION
**Pull request by amaid**
## Overview

Under media source, currently we only support local file and OneDrive, this improvement will add another media source i.e. IPFS media source.

Closes #242

Following the approach of OneDrive media source, the FileBrowser and FileSelector was reused to show the folders from ipfs.
Currently root folder selection using ipns address disabled.

## Checklist

This PR meets the following requirements:
- [x] Tested and contains **NO** breaking changes or known regressions.
- [x] Tested with the upstream branch merged in.
- [x] Tests have been added for bug fixes / features (or this option is not applicable)
- [x] All new code has been documented (or this option is not applicable)
- [x] Headers have been added to all new source files (or this option is not applicable)

## Additional info

Not provided

### CODE REVIEW

File: src/Shared/AppModels/AppRoot.cs [hunk 1]
```diff
@@ -245,6 +204,7 @@ public async Task InitAsync(CancellationToken cancellationToken = default)
                 return;
             }
 
+            Guard.IsNotNull(Ipfs);
```
Comment (Arlodotexe) [inline src/Shared/AppModels/AppRoot.cs:207]: This line needs to be removed. Ipfs is optional in Strix, this will crash if the user doesn't have IPFS enabled.

File: src/Shared/AppModels/AppRoot.cs [hunk 2]
```diff
@@ -373,6 +389,15 @@ private async Task HandleCoreSettingsCollectionChangedAsync<TSettings>(object? s
                         Logger.LogInformation($"Adding new core {newCore.DisplayName}, {nameof(newCore.InstanceId)} {newCore.InstanceId}");
                         _mergedCore.AddSource(newCore);
                     }
+
+                    Guard.IsNotNull(Ipfs);
```
Comment (Arlodotexe) [inline src/Shared/AppModels/AppRoot.cs:393]: This line needs to be removed. Ipfs is optional in Strix, this will crash if the user doesn't have IPFS enabled.

File: src/Shared/AppModels/AppRoot.cs [hunk 3]
```diff
@@ -373,6 +389,15 @@ private async Task HandleCoreSettingsCollectionChangedAsync<TSettings>(object? s
...
+                    var mergedCoreWithPlugins = CreatePluginLayer(_mergedCore, Ipfs);
+                    StrixDataRoot = new StrixDataRootViewModel(mergedCoreWithPlugins);
```
Comment (Arlodotexe) [inline src/Shared/AppModels/AppRoot.cs:395]: There's no need to do any of this new stuff. This method calls `InitAsync` again, as it has all the code to do this already and may do other things needed for setup. See [here](https://github.com/Arlodotexe/strix-music/blob/8ca4df16eba5a844704759ab8bf3475148487c45/src/Shared/AppModels/AppRoot.cs#L185).

File: src/Shared/AppModels/AppRoot.cs [hunk 4]
```diff
@@ -405,7 +430,7 @@ private async Task HandleCoreSettingsCollectionChangedAsync<TSettings>(object? s
             }
         }
 
-        await InitAsync();
+        await InitAsync(CancellationToken.None);
```
Comment (Arlodotexe) [inline src/Shared/AppModels/AppRoot.cs:433]: Redundant code change, this is equivalent to leaving the parameter empty.

File: src/Shared/Controls/Settings/MusicSources/ConnectNew/Ipfs/FolderSelector.xaml [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        Margin="20"
+        RowSpacing="20">
+        <Grid.RowDefinitions>
+            <RowDefinition Height="Auto" />
+            <RowDefinition Height="Auto" />
+            <RowDefinition Height="Auto" />
+            <RowDefinition Height="Auto" />
+        </Grid.RowDefinitions>
+
+        <!--<StackPanel Visibility="{x:Bind InvertBoolToVisibility(AuthenticateCommand.ExecutionTask.IsFaulted), Mode=OneWay, FallbackValue=Visible}">
```
Comment (Arlodotexe) [inline src/Shared/Controls/Settings/MusicSources/ConnectNew/Ipfs/FolderSelector.xaml:27]: This looks like leftover code, possibly copied from the OneDrive FolderSelector. Let's just remove this.

File: src/Shared/Controls/Settings/MusicSources/ConnectNew/Ipfs/LandingPage.xaml [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        MaxWidth="550"
+        Margin="20"
+        Spacing="15">
+
+        <labs:SettingsExpander
+            Description="Provide the root directory Cid."
+            Header="Configure IPFS folder"
+            IsEnabled="{x:Bind IsNotNull(Settings), Mode=OneWay}"
+            IsExpanded="True"
+            Visibility="{x:Bind NullToVisibility(_param.AppRoot.Ipfs.Client)}">
```
Comment (Arlodotexe) [inline src/Shared/Controls/Settings/MusicSources/ConnectNew/Ipfs/LandingPage.xaml:22]: We shouldn't bind to fields. Add an `[ObservableProperty]` to the field and use the generated property instead.

File: src/Shared/Controls/Storage/FolderBrowser.xaml.cs [hunk 7]
```diff
@@ -94,8 +130,15 @@ private async Task OnCurrentFolderChanged(IFolder? oldValue, IFolder? newValue)
         if (newValue is null)
             return;
 
-        await foreach (var item in newValue.GetItemsAsync())
+        FetchingItems = true;
+        IsFolderEmpty = false;
+
+        await foreach (var item in newValue.GetItemsAsync(StorableType.All))
```
Comment (Arlodotexe) [inline src/Shared/Controls/Storage/FolderBrowser.xaml.cs:136]: Passing the default value isn't needed.

```suggestion
        await foreach (var item in newValue.GetItemsAsync())
```

File: src/Shared/Controls/Storage/FolderBrowser.xaml [hunk 8]
```diff
@@ -30,20 +30,21 @@
                 ToolTipService.ToolTip="Navigate up"
                 Visibility="{x:Bind GoToParentCommand.IsRunning, Mode=OneWay}" />
 
-            <ProgressRing IsActive="{x:Bind GoToParentCommand.IsRunning, Mode=OneWay}" Visibility="{x:Bind GoToParentCommand.IsRunning, Mode=OneWay}" />
-
             <TextBlock Text="{x:Bind Name, Mode=OneWay}" />
+
+            <ProgressRing IsActive="{x:Bind FetchingItems, Mode=OneWay}" />
```
Comment (Arlodotexe) [inline src/Shared/Controls/Storage/FolderBrowser.xaml:35]: The XAML in this file was using commands and `x:Bind`-ing to data directly to avoid creating UI-specific properties like these. Why was this change needed?

File: src/Shared/Settings/IpfsCoreSettings.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using ColorCode.Common;
```
Comment (Arlodotexe) [inline src/Shared/Settings/IpfsCoreSettings.cs:5]:
```suggestion
```

File: src/Shared/Settings/IpfsCoreSettings.cs [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        public string IpnsAddress
+        {
+            get => GetSetting(() => string.Empty);
+            set => SetSetting(value);
+        }
+
+        /// <inheritdoc />
+        public override bool IsSettingValidForCoreCreation(string propertyName, object? value) => propertyName switch
+        {
+            nameof(InstanceId) or nameof(InstanceId) or nameof(IpfsCidPath) or nameof(IpfsCidPath)
```
Comment (Arlodotexe) [inline src/Shared/Settings/IpfsCoreSettings.cs:65]:
```suggestion
            nameof(InstanceId) or nameof(IpfsCidPath) or nameof(IpfsCidPath)
```

File: src/Shared/Settings/IpfsCoreSettings.cs [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+                => !string.IsNullOrWhiteSpace((string?)value ?? string.Empty),
+            _ => true,
+        };
+
+        /// <inheritdoc />
+        public override object GetSettingByName(string settingName)
+        {
+            Guard.ArgNotNull(settingName, nameof(settingName));
```
Comment (Arlodotexe) [inline src/Shared/Settings/IpfsCoreSettings.cs:73]:
```suggestion
            Guard.IsNotNull(settingName);
```

File: src/Shared/Settings/IpfsSettings.cs [hunk 12]
```diff
@@ -90,7 +90,7 @@ public bool GlobalPlaybackStateCountPluginEnabled
     /// </summary>
     public string ReleaseIpns
     {
-        get => GetSetting(() => Environment.GetEnvironmentVariable(nameof(ReleaseIpns)) ?? "/ipns/latest.strixmusic.com");
+        get => GetSetting(() => Environment.GetEnvironmentVariable(nameof(ReleaseIpns)) ?? "/ipns/staging.strixmusic.com");
```
Comment (Arlodotexe) [inline src/Shared/Settings/IpfsSettings.cs:93]:
```suggestion
        get => GetSetting(() => Environment.GetEnvironmentVariable(nameof(ReleaseIpns)) ?? "/ipns/latest.strixmusic.com");
```

File: src/Shared/Settings/LocalStorageCoreSettings.cs [hunk 13]
```diff
@@ -1,6 +1,7 @@
 ﻿using System;
 using OwlCore.ComponentModel;
 using OwlCore.Diagnostics;
+using OwlCore.Kubo;
```
Comment (Arlodotexe) [inline src/Shared/Settings/LocalStorageCoreSettings.cs:4]:
```suggestion
```

File: src/Shared/Settings/MusicSourcesSettings.cs [hunk 14]
```diff
@@ -52,6 +52,14 @@ public MusicSourcesSettings(IModifiableFolder folder)
             defaultSettingsFactory: async instanceId => new OneDriveCoreSettings(await GetDataFolderByName(instanceId.HashMD5Fast())))
         );
 
+        AvailableMusicSources.Add(new AvailableMusicSource
+            (
+            name: "IPFS",
+            description: "Stream music directly from ipfs folder",
```
Comment (Arlodotexe) [inline src/Shared/Settings/MusicSourcesSettings.cs:58]:
```suggestion
            description: "Play music stored on ipfs",
```

File: src/Shared/Controls/Settings/MusicSources/ConnectNew/Ipfs/FolderSelector.xaml [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+                FontSize="12"
+                Foreground="Red"
+                Opacity="0.5"
+                Text="{x:Bind AuthenticateCommand.ExecutionTask.Exception.Message, FallbackValue='', Mode=OneWay}"
+                TextAlignment="Center"
+                TextWrapping="WrapWholeWords" />
+        </StackPanel>-->
+
+        <StackPanel Grid.Row="1" Spacing="12">
+            <Image MaxWidth="250" Source="ms-appx:///Assets/Cores/OneDrive/login.svg" />
```
Comment (Arlodotexe) [inline src/Shared/Controls/Settings/MusicSources/ConnectNew/Ipfs/FolderSelector.xaml:63]: We should use a different svg for the Ipfs selector. Pick a good one from https://undraw.co/ with the color set to Strix purple: #BA27A9
```suggestion
            <Image MaxWidth="250" Source="ms-appx:///Assets/Cores/Ipfs/login.svg" />
```

Comment (amaid) [reply src/Shared/Controls/Settings/MusicSources/ConnectNew/Ipfs/FolderSelector.xaml]: Resolved with this icon:
image

File: src/Shared/AppModels/CoreFactory.cs [hunk 16]
```diff
@@ -141,4 +143,43 @@ public static async Task<StorageCore> CreateOneDriveCoreAsync(OneDriveCoreSettin
...
+
+        var coreDataFolder = await settings.Folder.GetFoldersAsync().FirstOrDefaultAsync(x => x.Name == instanceId) ??
+                             await settings.Folder.CreateFolderAsync(instanceId);
+
+        if (coreDataFolder is not IModifiableFolder modifiableCoreDataFolder)
+            throw new InvalidOperationException($"A new folder was created in the data folder, but it's not modifiable.");
+
+        var folderToScan = new IpfsFolder(settings.IpfsCidPath, client);
+
+        var logoFile = await StorageFile.GetFileFromApplicationUriAsync(new Uri("ms-appx:///Assets/Cores/LocalStorage/Logo.svg"));
```
[the remark this answers is not in the document]
Comment (amaid) [reply src/Shared/AppModels/CoreFactory.cs]: Nice catch

File: src/Shared/Controls/Storage/FolderBrowser.xaml (hunk 8 above)
Comment (amaid) [reply src/Shared/Controls/Storage/FolderBrowser.xaml:35]: I can't properly remember but there was major bug with existing implementation where there was no busy indicator for a long running task while fetching folders
image

here's the commit of that time:
https://github.com/Arlodotexe/strix-music/pull/278/commits/501c9651dbe681e941b7116ece35540488c13432

Comment (Arlodotexe) [reply src/Shared/Controls/Storage/FolderBrowser.xaml:35]: It's built into AsyncRelayCommand

Comment (Arlodotexe) [reply src/Shared/Controls/Storage/FolderBrowser.xaml:35]: Instead of `IsFolderEmpty`, use the existing `IsZeroToBoolean` function to bind `CurrentFolderItems.Count`. The fixup for `FetchingItems` is a bit more complex, I'll do it later.

Comment (amaid) [reply src/Shared/Controls/Storage/FolderBrowser.xaml:35]: Ah neat, removed IsFolderEmpty boolean and used InverseToVisibilityConverter on FetchingItems boolean, since we need to track the state of the "FetchingItems" we can't directly check CurrentFolderItems count which will be zero on control load as the content hasn't been fetched by then since it takes sometime. So the correct state of the folder Items count can only be determined after getting response from the ipfsClient.

**Comment by Arlodotexe** [comment]
Found a minor UI issue, "folder empty" label showing when it shouldn't.

image

**Comment by Arlodotexe** [comment]
I've also noted that DNSLink addresses don't work for IPNS? The `IpnsFolder` in OwlCore.Storage should be able to handle both of these:
- `/ipns/k51qzi5u..../`
- `/ipns/mymusic.mydomain.io/`

**Comment by Arlodotexe** [comment]
Upon adding this new music source, nothing happens:

https://github.com/Arlodotexe/strix-music/assets/9384894/0968e7eb-8a9f-40b0-a004-a6bcf048631e

File: src/Shared/Helpers/IpfsAddressValidator.cs [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+using System.Collections.Generic;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace StrixMusic.Helpers
+{
+    /// <summary>
+    /// Validates ipfs addresses.
+    /// </summary>
+    public class IpfsAddressValidator
```
Comment (Arlodotexe) [inline src/Shared/Helpers/IpfsAddressValidator.cs:11]: This isn't needed. `Ipns` addresses can also be treated as `Cid`s. the `Cid` type validates whatever you give it on construction. 

The code to check CIDs also limits you to CIDv0. We need to be supported CIDv1 and all possible multihashes (included libp2p public key hashes, the ones used for ipns).

Let's just use the `Cid` type provided by `Ipfs.Core`.

**Review by Arlodotexe — APPROVED** [review]

**Review by Arlodotexe — APPROVED** [review]
